### ISSUE TITLE
SDL_Mixer's Mix_HaltMusic freezes with certain non-default timidity.cfg option

### ISSUE THREAD

**Issue by ohrrpgce-bugbot** [ohrrpgce/ohrrpgce#659]
[bz#659]

Steps to reproduce:

Open the music importer.

Select a different song (either with arrow keys or numbers)

Select a third song song

The program will freeze before leaving the second song.

it doesn't matter which song numbers you use.

Only tested on Linux so far.

From: @bob-the-hamster
Reported version: 20110506 Zenzizenzic
Operating system:   Linux (Any)
Severity:     minor

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster

Ah. This is not our bug. I had been playing with my timidity.cfg file recently (in an attempt to get some better instruments) and I noticed the following:

\## If you have a moderate CPU, try these:
\#opt EFresamp=l
\#opt EFreverb=g,42
\#opt EFchorus=s
\#opt s32kHz
\#opt p64a

I un-commented those options, and indeed, it did result in a very slight improvement to sound quality when playing midi files using the timidity command-line tool.

However, next time I tried to use the Import Music menu in Wandering hamster, I get freezes. It seems that changing any one of these options in timidity.cfg causes SDl_mixer to freeze when it tries to stop a midi the second time.

Since this is either a Timidity bug or an SDL_Mixer, I am reducing the priority of this bug.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @rversteegen

\*\*\* Bug bz⁠#640 has been marked as a duplicate of this bug. \*\*\*

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @rversteegen

I'm seeing crashing in SDL_Mixer's timidity under both Windows (see bug bz⁠#640) and Linux. In both cases, it happens when I exit the music menu (music_close is called), and then reenter it (sdl_mixer is reinitialised).

Under Windows, I'm using the GUS patches linked to on the SDL_Mixer page, under Linux I'm using eawpats, with no changes to the config files. In fact, I think my timidity.cfg is nearly empty.

Though these might be different bugs. We could probably fix mine by only initialising SDL_mixer once.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster


> I'm seeing crashing in SDL_Mixer's timidity under both Windows (see bug bz⁠#640)
> and Linux. In both cases, it happens when I exit the music menu (music_close is
> called), and then reenter it (sdl_mixer is reinitialised).
> 
> Under Windows, I'm using the GUS patches linked to on the SDL_Mixer page, under
> Linux I'm using eawpats, with no changes to the config files. In fact, I think
> my timidity.cfg is nearly empty.
> 
> Though these might be different bugs. We could probably fix mine by only
> initialising SDL_mixer once.
> 

This sounds like a different bug to me. I agree that we definitely should only be initing SDL_Mixer once.... hmmm... although actually, I thought we already had code to prevent double-init...

		if sdl_init_done = 0 then
			SDL_Init(SDL_INIT_AUDIO)
			sdl_init_done = -1
			sdl_audio_only = -1
		end if

Anyway, in custom we only init music when entering the importer, which is a throwback to the old DO days when i was afraid about memory usage from the music buffer. I think we should just init once when custom first starts, just like we do in game.

As an added bonus, we can make music and sound previewable everywhere that makes sense. I have filed bug bz⁠#661 for that

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @rversteegen


> This sounds like a different bug to me. I agree that we definitely should only
> be initing SDL_Mixer once.... hmmm... although actually, I thought we already
> had code to prevent double-init...

Yes, but I'm saying that SDL_Mixer apparently can't be reinitialised afterwards.

**Comment by rversteegen** [comment]
I added a new "resolved: obsolete" label for unfixed bugs pertaining to obsolete platforms/configurations.  We have a pile of these old 2008-2009 MIDI on Ubuntu bugs. The fork of Timidity in SDL_mixer has received fixes over the years (and it's completely replaced with a different fork of TiMidity++ in the unreleased SDL_mixer 1.2.13). I tried to reproduce this bug using SDL_mixer 1.2.12 (using timidity) and found that adding any of those lines to timidity.cfg (which are the wrong syntax anyway according to my `man timidity.cfg`) caused SDL_mixer to simply not play MIDI files.
